---
title: "HPRT1 and TK1 transcript expressions across cell lines"
output: html_document
---
### Parsing Data
The dataset is based on the Cancer Cell Line Encycolpedia (CCLE) and can be found [here](https://osf.io/gqrz9). Go to matrices -> CCLE -> CCLE_tpm.tsv.gz.
```{r setup, include=FALSE}
library(rmarkdown)
library(tidyverse)
library(ggplot2)
library(dplyr)
require(readr)
library(knitr)
```

A python script (separate from this .Rmd script) was created to filter the CCLE data based on desired cell lines of genes HPRT1 and TK1. The cell lines are part of the following groups:

* Wild type
* Null
* Codon 175 mutations (R175H)
* Codon 248 mutations (R248Q)
* Codon 273 mutations (R273H)

All transcript expression values are log2-transformed to mitigate the effects of outliers. 1 is also added to prevent errors in case there are any zero values. 
```{r, message=FALSE, warning=FALSE}
# Read trimmed data as dataframe
CCLE_data <- as.data.frame(read_tsv("./transcripts_HPRT1_TK1.tsv")) 

# log2 transform transcript expression values
CCLE_data[, "Value"] <- log2(CCLE_data[, "Value"] + 1)

# Function that filters data by transcript ID
filter_by_ID <- function(df, transcriptID) {
  filtered_data <- filter(df, TranscriptID == transcriptID)
  return(filtered_data)
}

# Create new dataframes based on gene and transcript ID
# HPRT1
HPRT1 <- filter(CCLE_data, Gene == "HPRT1")
transcripts_HPRT1 <- unique(HPRT1$TranscriptID)  # Get transcript IDs in HPRT1

for (transcript in transcripts_HPRT1) {
  new_df <- gsub("-", "_", transcript)
  assign(new_df, filter_by_ID(HPRT1, transcript))
}

# TK1
TK1 <- filter(CCLE_data, Gene == "TK1")
transcripts_TK1 <- unique(TK1$TranscriptID)  # Get transcript IDs in TK1

for (transcript in transcripts_TK1) {
  new_df <- gsub("-", "_", transcript)
  assign(new_df, filter_by_ID(TK1, transcript))
}
```

### ANOVA and Tukey Tests

Multiple comparison analysis is performed using [Tukey's test](https://en.wikipedia.org/wiki/Tukey%27s_range_test).

In each gene...

1. We analyze the means between each transcript ID.
2. Then, we analyze the means between each cell line group.

```{r, message=FALSE, warning=FALSE}
# HPRT1
# Comparing means between each transcript ID
HPRT1_av <- aov(Value ~ TranscriptID, data = HPRT1)
HPRT1_tukey <- TukeyHSD(HPRT1_av)

# Comparing means between each group in each transcript ID
HPRT1_001_av <- aov(Value ~ Group, data = HPRT1_001)
HPRT1_001_tukey <- TukeyHSD(HPRT1_001_av)

HPRT1_002_av <- aov(Value ~ Group, data = HPRT1_002)
HPRT1_002_tukey <- TukeyHSD(HPRT1_002_av)

HPRT1_003_av <- aov(Value ~ Group, data = HPRT1_003)
HPRT1_003_tukey <- TukeyHSD(HPRT1_003_av)

# TK1
# Comparing means between each transcript ID
TK1_av <- aov(Value ~ TranscriptID, data = TK1)
TK1_tukey <- TukeyHSD(TK1_av)

# Comparing means between each group in each transcript ID
TK1_001_av <- aov(Value ~ Group, data = TK1_001)
TK1_001_tukey <- TukeyHSD(TK1_001_av)

TK1_002_av <- aov(Value ~ Group, data = TK1_002)
TK1_002_tukey <- TukeyHSD(TK1_002_av)

TK1_003_av <- aov(Value ~ Group, data = TK1_003)
TK1_003_tukey <- TukeyHSD(TK1_003_av)

TK1_004_av <- aov(Value ~ Group, data = TK1_004)
TK1_004_tukey <- TukeyHSD(TK1_004_av)

TK1_005_av <- aov(Value ~ Group, data = TK1_005)
TK1_005_tukey <- TukeyHSD(TK1_005_av)

TK1_006_av <- aov(Value ~ Group, data = TK1_006)
TK1_006_tukey <- TukeyHSD(TK1_006_av)
```

All p-values from the statistical analyses above are placed in the following files, separated by gene:

* "HPRT1_transcript_pvals.tsv"
* "TK1_transcript_pvals.tsv"
```{r, message=FALSE, warning=FALSE}
options(scipen = 999) # Keep values in numeric form instead of scientific notation

# Function that creates a dataframe of tukey p-values
# Dataframes needed to print p-values to a file
make_dataframe <- function(tukey_name, transcript) {
  df_name <- data.frame(tukey_name$Group)
  df_name <- cbind(rownames(df_name), data.frame(df_name, row.names = NULL))
  colnames(df_name)[1] <- "Group"
  colnames(df_name)[5] <- transcript
  df_name <- df_name[, c("Group", transcript)]
  return(df_name)
}

# Make a dataframe for each HPRT1 transcript
HPRT1_001_df <- make_dataframe(HPRT1_001_tukey, "HPRT1-001")
HPRT1_002_df <- make_dataframe(HPRT1_002_tukey, "HPRT1-002")
HPRT1_003_df <- make_dataframe(HPRT1_003_tukey, "HPRT1-003")

# Merge all HPRT1 dataframes and write to a file
HPRT1_pvals <- list(HPRT1_001_df, HPRT1_002_df, HPRT1_003_df) %>% reduce(full_join, by = "Group")
write.table(HPRT1_pvals, file = "./HPRT1_transcript_pvals.tsv", sep = "\t", quote = FALSE, row.names = FALSE)

# Make a dataframe for each TK1 transcript
TK1_001_df <- make_dataframe(TK1_001_tukey, "TK1-001")
TK1_002_df <- make_dataframe(TK1_002_tukey, "TK1-002")
TK1_003_df <- make_dataframe(TK1_003_tukey, "TK1-003")
TK1_004_df <- make_dataframe(TK1_004_tukey, "TK1-004")
TK1_005_df <- make_dataframe(TK1_005_tukey, "TK1-005")
TK1_006_df <- make_dataframe(TK1_006_tukey, "TK1-006")

# Merge all TK1 dataframes and write to a file
TK1_pvals <- list(TK1_001_df, TK1_002_df, TK1_003_df, TK1_004_df, TK1_005_df, TK1_006_df) %>% reduce(full_join, by = "Group")
write.table(TK1_pvals, file = "./TK1_transcript_pvals.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
```

### Plots
The boxplots outputted by the code below illustrates transcript expression in HPRT1 and TK1. The graphs are organized by transcript ID, as seen in the labels of the grey boxes. In addition, the following cell lines are highlighted:

* DU145 (from R273H)
* PC3 (from Null)
```{r, message=FALSE, warning=FALSE}
make_boxplot_combined <- function(dataframe, gene) {
  set.seed(101)
  ggplot(dataframe, aes(x = Group, y = Value, fill = Group)) +
  geom_boxplot(outlier.shape = NA, alpha = 0.5) +
  geom_jitter(alpha = 0.3) +
  geom_jitter(data = filter(dataframe, Cell_Line == "DU_145"), aes(colour = "DU_145"), alpha = 0.9) +
  geom_jitter(data = filter(dataframe, Cell_Line == "PC-3"), aes(colour = "PC-3"), alpha = 0.9) +
  scale_colour_hue(name = "Cell Line", l = 50) +
  scale_y_continuous(name = "Transcript-Level Value\n(log2 transformed)",
                     breaks = seq(0,10,1.5)) +
  scale_x_discrete(name = "") +
  ggtitle(paste("Transcript Expression in", gene)) +
  guides(fill = guide_legend(title = "Cell Line Group")) +
  theme_bw() +
  theme(plot.title = element_text(size = 14, face = "bold", hjust = 0.5),
        axis.title = element_text(face = "bold"),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank()) +
  facet_grid(. ~ TranscriptID)
}

print(make_boxplot_combined(HPRT1, "HPRT1"))
ggsave(file="HPRT1_all.pdf", width=8)
print(make_boxplot_combined(TK1, "TK1"))
ggsave(file="TK1_all.pdf", width=12)
```

The boxplots outputted by the code below contain exactly the same data as the boxplots above. Here, however, each graph only contains data from a single transcript ID. 
```{r, message=FALSE, warning=FALSE}
make_boxplot_ind <- function(dataframe, transcript, type, break_size) {
  set.seed(101)
  ggplot(dataframe, aes(x = Group, y = Value, fill = Group)) +
  geom_boxplot(outlier.shape = NA, alpha = 0.5) +
  geom_jitter(alpha=0.3) +
  geom_jitter(data = filter(dataframe, Cell_Line == "DU_145"), aes(colour = "DU_145"), alpha = 0.9) +
  geom_jitter(data = filter(dataframe, Cell_Line == "PC-3"), aes(colour = "PC-3"), alpha = 0.9) +
  scale_colour_hue(name = "Cell Line", l = 40) +
  scale_y_continuous(name = "Transcript-Level Value\n(log2 transformed)",
                     breaks = seq(0,10,break_size)) +
  scale_x_discrete(name = "") +
  ggtitle(paste(transcript, " Transcript Expression (", type, ")", sep = "")) +
  guides(fill = guide_legend(title = "Cell Line Group")) +
  theme_bw() +
  theme(plot.title = element_text(size = 14, face = "bold", hjust = 0.5),
        axis.title = element_text(face = "bold"),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank())
}

print(make_boxplot_ind(HPRT1_001, "HPRT1-001", "Protein Coding", 0.5))
ggsave(file="HPRT1_001.pdf", width=6.5)
print(make_boxplot_ind(HPRT1_002, "HPRT1-002", "Processed Transcript", 0.25))
ggsave(file="HPRT1_002.pdf", width=6.5)
print(make_boxplot_ind(HPRT1_003, "HPRT1-003", "Processed Transcript", 0.5))
ggsave(file="HPRT1_003.pdf", width=6.5)

print(make_boxplot_ind(TK1_001, "TK1-001", "Protein Coding", 0.5))
ggsave(file="TK1_001.pdf", width=6.5)
print(make_boxplot_ind(TK1_002, "TK1-002", "Protein Coding", 0.5))
ggsave(file="TK1_002.pdf", width=6.5)
print(make_boxplot_ind(TK1_003, "TK1-003", "Retained Intron", 0.5))
ggsave(file="TK1_003.pdf", width=6.5)
print(make_boxplot_ind(TK1_004, "TK1-004", "Protein Coding", 0.5))
ggsave(file="TK1_004.pdf", width=6.5)
print(make_boxplot_ind(TK1_005, "TK1-005", "Protein Coding", 0.5))
ggsave(file="TK1_005.pdf", width=6.5)
print(make_boxplot_ind(TK1_006, "TK1-006", "Protein Coding", 1.5))
ggsave(file="TK1_006.pdf", width=6.5)
```

